File: git_repo_searcher_flutter/lib/domain/sign_in_use_case.dart
import 'package:dio/dio.dart';
import 'package:gitreposearcherflutter/data/client_keys_provider.dart';
import 'package:gitreposearcherflutter/data/graphql_client_provider.dart';
import 'package:gitreposearcherflutter/data/login_prefs.dart';
import 'package:gitreposearcherflutter/data/main_dio_provider.dart';
import 'package:gitreposearcherflutter/domain/use_case.dart';
import 'package:gitreposearcherflutter/ui/strings.dart';

class SignInUseCase extends UseCase<String, bool> {
  final Dio _service;
  final LoginPrefs _loginPrefs;
  final MainDioProvider _dioProvider;
  final GraphqlClientProvider _graphProvider;
  final ClientKeyProvider _clientKeyProvider;
  final _url = "https://github.com/login/oauth/access_token";

  SignInUseCase(this._service, this._loginPrefs, this._dioProvider,
      this._graphProvider, this._clientKeyProvider);

  @override
  Future<bool> execute([String params]) async {
    final clientKeyModel = await _clientKeyProvider.loadKeys();
    final response = await _service.post(_url, queryParameters: {
      "code": params,
      "client_id": clientKeyModel.clientId,
      "client_secret": clientKeyModel.clientSecret
    });
    final accessToken = response.data
        .toString()
        .split("&")
        .firstWhere((element) => element.contains("access_token"),
            orElse: () => null)
        ?.split("=")
        ?.last;
    if (accessToken == null) throw Exception("null token received");
    _loginPrefs.setToken(accessToken);
    _dioProvider.setToken(accessToken);
    _graphProvider.setToken(accessToken);
    return true;
  }
}

File: git_repo_searcher_kotlin/app/src/main/java/com/example/gitreposearcher/domain/SignInUseCase.kt
package com.example.gitreposearcher.domain

import android.content.Context
import com.example.gitreposearcher.R
import com.example.gitreposearcher.checkResponse
import com.example.gitreposearcher.data.git_api.TokenInterceptor
import com.example.gitreposearcher.data.shrared_prefs.LoginPrefs
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request

class SignInUseCase(
    private val ctx: Context,
    private val client: OkHttpClient,
    private val loginPrefs: LoginPrefs,
    private val tokenInterceptor: TokenInterceptor
) {
    suspend operator fun invoke(code: String) {
        val url = "https://github.com/login/oauth/access_token".toHttpUrlOrNull() ?: return
        val requestBody = FormBody.Builder()
            .add("code", code)
            .add("client_id", ctx.resources.getString(R.string.client_id))
            .add("client_secret", ctx.resources.getString(R.string.client_secret))
            .build()
        val tokenObject = withContext(Dispatchers.IO) {
            val response =
                client.newCall(Request.Builder().url(url).post(requestBody).build()).execute()
            response.checkResponse()
            val tokenData = response.body?.string() ?: throw Throwable("null response")
            val tokenKeyRegex = Regex("access_token")
            val accessToken =
                tokenData.split("&").firstOrNull { it.contains(tokenKeyRegex) }?.split("=")?.last()
                    ?: throw Throwable("token not found")
            // val result = Gson().fromJson(tokenData, TokenApiModel::class.java)
            with(loginPrefs) {
                token = accessToken
            }
            accessToken
        }
        tokenInterceptor.token = tokenObject
    }
}
